functionary/ui/views/schedules/update.py

```diff
@@ -0,0 +1,85 @@
+from django.http import HttpRequest, HttpResponse, HttpResponseRedirect, QueryDict
+from django.shortcuts import get_object_or_404, render
+from django.urls import reverse
+
+from core.auth import Permission
+from core.models import Environment, Function, ScheduledTask
+from ui.forms import ScheduledTaskForm, TaskParameterForm
+from ui.views.schedules.utils import get_crontab_schedule
+from ui.views.view_base import PermissionedFormUpdateView
+
+
+class ScheduledTaskUpdateView(PermissionedFormUpdateView):
+    model = ScheduledTask
+    form_class = ScheduledTaskForm
+    template_name = "forms/schedules/scheduling_create_or_update.html"
+
+    def get_success_url(self) -> str:
+        return reverse("ui:schedule-list")
+
+    def get_initial(self) -> dict:
+        initial = super().get_initial()
+        scheduled_task: ScheduledTask = self.get_object()
+        crontab = scheduled_task.periodic_task.crontab
+        initial["scheduled_minute"] = crontab.minute
+        initial["scheduled_hour"] = crontab.hour
+        initial["scheduled_day_of_week"] = crontab.day_of_week
+        initial["scheduled_day_of_month"] = crontab.day_of_month
+        initial["scheduled_month_of_year"] = crontab.month_of_year
+        return initial
+
+    def get_context_data(self, **kwargs) -> dict:
+        context = super().get_context_data(**kwargs)
+        scheduled_task: ScheduledTask = self.get_object()
+        context["update"] = True
+        context["task_parameter_form"] = TaskParameterForm(
+            function=scheduled_task.function, initial=scheduled_task.parameters
+        )
+        return context
+
+    def get_form_kwargs(self) -> dict:
+        kwargs = super().get_form_kwargs()
+        environment = get_object_or_404(
+            Environment, id=self.request.session.get("environment_id")
+        )
+        kwargs["environment"] = environment
+        return kwargs
+
+    def post(self, request: HttpRequest, **kwargs) -> HttpResponse:
+        scheduled_task: ScheduledTask = self.get_object()
+        data: QueryDict = request.POST.copy()
+        data["environment"] = scheduled_task.environment
```

Comment:
Two things here:

* You could create a separate `ScheduledTaskUpdateForm` that excludes the environment field, since you don't want that to change.  Then you wouldn't worry about explicitly setting it here.  See the `WorkflowStepCreateForm` and `WorkflowStepUpdateForm` as an example, which does what I'm describing with the "workflow" field, since that should never be updated after initial creation.

* I'm kind of of the opinion that the function shouldn't change on a scheduled task after initial creation either.  As soon as you change the function it is for all intents an purposes an entirely new scheduled task, just weirdly associated with the tasking history of some other scheduled task.  If we were tracking the revision history of the scheduled tasks I could see maybe allowing it, but for now I'd vote to not.

Reply:
Since the update view already has context of the object its editing, I just stamp in the environment and the function from the current object, which basically makes those fields of the ScheduledTask immutable. I disabled the function field so that the user knows they shouldn't be editing the function.